File: sdk.py
from subprocess import run
from functools import reduce

delimiter = ":"
fileEndings = {'javascript':'js'}

def meta(code:str, lang:str, input:dict, output:list):
    #1) Make codefile
    codeFile = open("code." + fileEndings[lang], 'w')
    codeFile.write(code)
    codeFile.close()
    #2) Make transfer file
    makeTransferFile(input)
    #3) Run execute with run.py
    if output:
        #Can't reduce an empty array
        outputArgs = reduce(lambda a,b:a+b+" ",output)
    else:
        outputArgs = ""
    print(run("python3 run.py transfer.txt python3 %s code.%s %s" % (lang, fileEndings[lang], outputArgs), shell=True))
    #4) Parse output
    return parse(lang, "transfer.txt")

def convert(lang:str, typ:str, data:str):
    if lang == "javascript":
        if typ == "string":
            return data
        elif typ == "number":
            if "." in data:
                return float(data)
            else:
                return int(data)

def parse(lang:str, filename:str):
    #Takes input file and outputs dictionary of converted variables
    variables = {}

    input = open(filename, 'r')
    while True:
        line = input.readline()
        if not line:
            break

        components = line.split(delimiter)
        name = components[0]
        typ = components[1]
        data = components[2]
        if not name == '': #Don't make a ghost variable for extra blank lines
            variables[name] = convert(lang, typ, data)
    
    return variables #Need to figure out how to get this into the globals() of code.py

def makeTransferFile(variables):
    #variables is a dictionary
    out = open("transfer.txt", 'w')
    i=0
    for key in variables.keys():
        val = variables[key]
        if i != 0:
            out.write("\n")
        out.write(key + ":" + str(type(val)).strip("<class ").strip("'>") + ":" + str(val))
        i+=1
    out.close()

File: sdk.js
const exec = require('child_process').exec;
const fs = require('fs');

const fileEndings = {'python3': 'py'};
const fileDelimiter = ":";

function makeTransferFile(input) {
    //Input is a dictionary
    var data = '';
    const keys = Object.keys(input);
    for (var i=0;i<keys.length;i++) {
        const key = keys[i];
        const val = input[key];
        const type = typeof val;
        const name = key;
        if (i != 0) {
            data += "\n";
        }
        data += name + fileDelimiter + type + fileDelimiter + String(val);
    }
    //transfer.txt will be overwritten each time because it isn't needed any longer than its first use
    fs.writeFile("transfer.txt", data, (err) => {
        if (err) throw "Error writing to input file: " + err.message;
    });
}

async function meta(code, lang, input, output) {

    const codeFile = fs.writeFile("code." + fileEndings[lang], code, (err) => {
        if (err) throw "Error writing to input file: " + err.message;
    });
    //input is a dictionary
    makeTransferFile(input);

    const promise = new Promise(function(resolve, reject) {
        exec(`python3 run.py input.txt javascript ${lang} code.${fileEndings[lang]} ${output}`, (err, stdout, stderr) => {
            variables = parse(lang, "transfer.txt");
            resolve(variables);
        });
    });
    return promise;
}

function convert(lang, typ, data) {
    //Data comes in Javascript String form
    switch (lang) {
        case "python3":
            switch (typ) {
                case "str":
                    return data;
                case "int":
                    return parseInt(data);
                default:
                    return data;
            }
    
        default:
            console.log("Unsupported language: " + lang);
    }
}

function parse(lang, filename) {
    //Takes input file and outputs dictionary of converted variables
    var variables = {};

    const lines = fs.readFileSync('transfer.txt', 'utf8').split('\n');
    for (line of lines) {
        //Process line of output file
        const comps = line.split(fileDelimiter);
        if (comps[0] != '') {
            //Don't make a ghost variable for extra blank lines
            variables[comps[0]] = convert(lang, comps[1], comps[2]);
        }
    }
    return variables;
}

exports.meta = meta;
exports.parse = parse;
exports.makeTransferFile = makeTransferFile;
exports.convert = convert;